File: src/main/scala/clase6/ADTBundledDeclarativeSecure.scala
package clase6

object ADTBundledDeclarativeSecure extends App {

  case class Stack[T](push: T => Stack[T], pop: () => (T, Stack[T]), isEmpty: () => Boolean)

  def stack[T]():Stack[T] = {

    def stackOps(content: List[T]): Stack[T] = {
      def push(element:T) = {
        stackOps(element :: content)
      }

      def pop() = {
        content match {
          case head :: tail =>
            (head, stackOps(tail))
        }
      }

      def isEmpty() = content == Nil

      Stack(push, pop, isEmpty)

    }

    stackOps(Nil)
  }

  assert(stack().isEmpty())

  val s = stack()
    .push("ariel")
    .push("alejandro")

  assert(!s.isEmpty())

  assert(s.pop()._1 == "alejandro")
  assert(s.pop()._2.pop()._1 == "ariel")
  assert(s.pop()._2.pop()._2.isEmpty())


}

File: haskell/src/clase6/ADTBundledDeclarativeSecure.hs
module ADTBundledDeclarativeSecure where

data Stack a = Stack {
  push :: a -> Stack a,
  pop :: (a, Stack a),
  isEmpty :: Bool
}

push_ content element = stackOps (element : content)
pop_ (head : tail) = (head, stackOps tail)

isEmpty_ [] = True
isEmpty_ content = False

stackOps content =
  Stack
    (push_ content)
    (pop_ content)
    (isEmpty_ content)

stack = stackOps []

main = let s = push (push stack "ariel") "alejandro" in do
    print (isEmpty stack) -- True
    print (isEmpty s) -- False
    print (fst (pop (s))) -- "alejandro"
    print (fst (pop (snd (pop (s))))) -- "ariel"
    print (isEmpty (snd (pop (snd (pop (s)))))) -- True
